homeassistant/components/plaato/__init__.py

```diff
@@ -59,32 +82,136 @@
+    """Init webhook based on config entry."""
+    if entry.data[CONF_WEBHOOK_ID] is not None:
+        webhook_id = entry.data.get(CONF_WEBHOOK_ID)
+        device_name = entry.data.get(CONF_DEVICE_NAME)
+        device_type = entry.data.get(CONF_DEVICE_TYPE)
+        device = {DEVICE_NAME: device_name, DEVICE_TYPE: device_type}
+        undo_listener = entry.add_update_listener(_async_update_listener)
+
+        hass.data[DOMAIN][entry.entry_id] = {
+            COORDINATOR: None,
+            DEVICE: device,
+            SENSOR_DATA: None,
+            UNDO_UPDATE_LISTENER: undo_listener,
+        }
+
+        hass.components.webhook.async_register(
+            DOMAIN, f"{DOMAIN}.{device_name}", webhook_id, handle_webhook
+        )
+    else:
+        raise InvalidStateError
+
+
+async def async_setup_coordinator(hass: HomeAssistant, entry: ConfigEntry):
+    """Init auth token based on config entry."""
+    auth_token = entry.data.get(CONF_TOKEN)
+    device_type = entry.data.get(CONF_DEVICE_TYPE)
+    undo_listener = entry.add_update_listener(_async_update_listener)
+    device = {
+        DEVICE_NAME: entry.data.get(CONF_DEVICE_NAME),
+        DEVICE_TYPE: device_type,
+        DEVICE_ID: auth_token,
+    }
+
+    coordinator = PlaatoCoordinator(hass, auth_token, device_type)
+    await coordinator.async_refresh()
+    if not coordinator.last_update_success:
+        raise ConfigEntryNotReady
+
+    hass.data[DOMAIN][entry.entry_id] = {
+        COORDINATOR: coordinator,
+        DEVICE: device,
+        SENSOR_DATA: None,
+        UNDO_UPDATE_LISTENER: undo_listener,
+    }
+
+    for platform in PLATFORMS:
+        if entry.options.get(platform, True):
+            coordinator.platforms.append(platform)
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
     """Unload a config entry."""
-    hass.components.webhook.async_unregister(entry.data[CONF_WEBHOOK_ID])
-    hass.data[SENSOR_DATA_KEY]()
+    use_webhook = entry.data.get(CONF_USE_WEBHOOK)
+    hass.data[DOMAIN][entry.entry_id][UNDO_UPDATE_LISTENER]()
+
+    if use_webhook:
+        return await async_unload_webhook(hass, entry)
+
+    return await async_unload_coordinator(hass, entry)
 
-    await hass.config_entries.async_forward_entry_unload(entry, SENSOR)
-    return True
+
+async def async_unload_webhook(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload webhook based entry."""
+    if entry.data[CONF_WEBHOOK_ID] is not None:
+        hass.components.webhook.async_unregister(entry.data[CONF_WEBHOOK_ID])
+
+    unloaded = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, platform)
+                for platform in PLATFORMS
+            ]
+        )
+    )
+    if unloaded:
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    return unloaded
+
+
+async def async_unload_coordinator(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload auth token based entry."""
```

Comment:
This is quite similar to `async_unload_webhook`. Please reduce duplicate code where possible